server/src/main/java/com/cloud/template/TemplateManagerImpl.java

```diff
@@ -1541,14 +1539,23 @@ public boolean updateTemplateOrIsoPermissions(BaseUpdateTemplateOrIsoPermissions
         //Derive the domain id from the template owner as updateTemplatePermissions is not cross domain operation
         Account owner = _accountMgr.getAccount(ownerId);
         final Domain domain = _domainDao.findById(owner.getDomainId());
+        final boolean isCallerRootAdmin = _accountMgr.isRootAdmin(caller.getAccountId());
         if ("add".equalsIgnoreCase(operation)) {
             final List<String> accountNamesFinal = accountNames;
             final List<Long> accountIds = new ArrayList<Long>();
             Transaction.execute(new TransactionCallbackNoReturn() {
                 @Override
                 public void doInTransactionWithoutResult(TransactionStatus status) {
                     for (String accountName : accountNamesFinal) {
-                        Account permittedAccount = _accountDao.findActiveAccount(accountName, domain.getId());
+                        Account permittedAccount = null;
+                        if (isCallerRootAdmin) {
+                            List<AccountVO> accountsWthName = _accountDao.findAccountsLike(accountName);
+                            if (accountsWthName.size() > 0) {
+                                permittedAccount = accountsWthName.get(0);
```

Comment:
@anuragaw could this cause regression or issues, this is always picking up the first account in the list; should it be something like - if user is root admin, then all enable/permit for all accounts matching the name using something like `accountIds.add(permittedAccount.getId());` in the if block; and refactor suitably?

Reply:
@rhtyd - good call. I looked at the implementation in more depth on Friday and observed the following -
1. Update template permissions uses account names (NOT uuids) to update permissions.
2. This means that the above API call is meant for domain specific sharing and hence implies a cross domain sharing restriction.
3. Double checked and confirmed that sharing templates VIA UI should also have same restrictions as updateTemplatePermissions API.

Hence I've reverted some changes and updated UI code to reflect these restrictions.